File: client/android/app/src/main/java/com/unity3d/player/FlutterUnityActivity.java
package com.unity3d.player;

import android.content.res.Configuration;
import android.os.Bundle;
import android.view.KeyEvent;
import android.view.MotionEvent;

import io.flutter.embedding.android.FlutterActivity;

public class FlutterUnityActivity extends FlutterActivity implements IUnityPlayerLifecycleEvents {

    public FlutterUnityPlayer mUnityPlayer;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mUnityPlayer = new FlutterUnityPlayer(this, this);
    }

    @Override
    protected void onDestroy() {
        mUnityPlayer.destroy();
        super.onDestroy();
    }

    @Override
    protected void onStop() {
        super.onStop();
        mUnityPlayer.stop();
    }

    @Override
    protected void onStart() {
        super.onStart();
        mUnityPlayer.start();
    }

    @Override
    protected void onPause() {
        super.onPause();
        mUnityPlayer.pause();
    }

    @Override
    protected void onResume() {
        super.onResume();
        mUnityPlayer.resume();
    }

    @Override
    public void onUnityPlayerUnloaded() {
        moveTaskToBack(true);
    }

    @Override
    public void onUnityPlayerQuitted() {}

    @Override
    public void onWindowFocusChanged(boolean hasFocus) {
        super.onWindowFocusChanged(hasFocus);
        mUnityPlayer.focusChanged(hasFocus);
    }

    @Override
    public void onConfigurationChanged(Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
        mUnityPlayer.configurationChanged(newConfig);
    }

    @Override
    public boolean dispatchKeyEvent(KeyEvent event) {
        if (event.getAction() == KeyEvent.ACTION_MULTIPLE) {
            return mUnityPlayer.injectEvent(event);
        }
        return super.dispatchKeyEvent(event);
    }

    @Override
    public boolean onKeyUp(int keyCode, KeyEvent event) {
        return mUnityPlayer.onKeyUp(keyCode, event);
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        return mUnityPlayer.onKeyDown(keyCode, event);
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        return mUnityPlayer.onTouchEvent(event);
    }

    @Override
    public boolean onGenericMotionEvent(MotionEvent event) {
        return mUnityPlayer.onGenericMotionEvent(event);
    }
}

File: client/android/app/src/main/java/com/unity3d/player/FlutterUnityActivity.kt
package com.unity3d.player

import android.content.res.Configuration
import android.os.Bundle
import android.view.KeyEvent
import android.view.MotionEvent
import io.flutter.embedding.android.FlutterActivity

open class FlutterUnityActivity : FlutterActivity(), IUnityPlayerLifecycleEvents {

    var mUnityPlayer: FlutterUnityPlayer? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mUnityPlayer = FlutterUnityPlayer(this, this)
    }

    override fun onDestroy() {
        mUnityPlayer?.destroy()
        super.onDestroy()
    }

    override fun onStop() {
        super.onStop()
        mUnityPlayer?.stop()
    }

    override fun onStart() {
        super.onStart()
        mUnityPlayer?.start()
    }

    override fun onPause() {
        super.onPause()
        mUnityPlayer?.pause()
    }

    override fun onResume() {
        super.onResume()
        mUnityPlayer?.resume()
    }

    override fun onUnityPlayerUnloaded() {
        moveTaskToBack(true)
    }

    override fun onUnityPlayerQuitted() {}

    override fun onWindowFocusChanged(hasFocus: Boolean) {
        super.onWindowFocusChanged(hasFocus)
        mUnityPlayer?.focusChanged(hasFocus)
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        mUnityPlayer?.configurationChanged(newConfig)
    }

    override fun dispatchKeyEvent(event: KeyEvent): Boolean {
        if (event.action == KeyEvent.ACTION_MULTIPLE) {
            return mUnityPlayer?.injectEvent(event) ?: false
        }
        return super.dispatchKeyEvent(event)
    }

    override fun onKeyUp(keyCode: Int, event: KeyEvent): Boolean =
        mUnityPlayer?.onKeyUp(keyCode, event) ?: false

    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean =
        mUnityPlayer?.onKeyDown(keyCode, event) ?: false

    override fun onTouchEvent(event: MotionEvent): Boolean =
        mUnityPlayer?.onTouchEvent(event) ?: false

    override fun onGenericMotionEvent(event: MotionEvent): Boolean =
        mUnityPlayer?.onGenericMotionEvent(event) ?: false
}
